### ISSUE TITLE
The punycode module is deprecated with Node.js 21.x

### ISSUE THREAD

**Issue by microaaron** [x3dom/x3dom#1316]
```
>npm run build

> x3dom@1.8.4-dev build
> node ./build/src-builder.js

> Recreate dist folder ...
> Concat files ...
> Minify x3dom.js ...
> Minify x3dom-full.js ...
> Minify x3dom-physics.js ...
> Write x3dom.debug.js ...
> Write x3dom.js ...
> Write x3dom-full.debug.js ...
> Write x3dom-full.js ...
> Write x3dom-physics.debug.js ...
> Write x3dom-physics.js ...
> Write VERSION ...
> Copy additional files ...
> Build documentation ...
(node:31800) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
> Build archive ...
```

**Comment by andreasplesch** [comment]
Thanks. Do you know of an alternative ?

**Comment by microaaron** [comment]
> Thanks. Do you know of an alternative ?

I have no idea.

**Comment by microaaron** [comment]
@andreasplesch By the way, a question.
https://github.com/x3dom/x3dom/blob/a0601f816d72b16309976ad5657e7212e4c26a5e/src/gfx_webgl.js#L703-L705
Why ```q6 = 6 * q;``` ? What if q>0?  The state after the for loop seems a little weird.
![x3d](https://github.com/x3dom/x3dom/assets/16946786/7caae962-31fa-4c66-bbcb-b1b462189dca)

https://github.com/x3dom/x3dom/blob/a0601f816d72b16309976ad5657e7212e4c26a5e/src/gfx_webgl.js#L842-L846
And, does it should be ```gl.enableVertexAttribArray( sp.binormal); ```?

**Comment by andreasplesch** [comment]
Great graphics !

> @andreasplesch By the way, a question.
> 
> https://github.com/x3dom/x3dom/blob/a0601f816d72b16309976ad5657e7212e4c26a5e/src/gfx_webgl.js#L703-L705
> 
> 
> Why `q6 = 6 * q;` ? What if q>0? The state after the for loop seems a little weird.
> ![x3d](https://private-user-images.githubusercontent.com/16946786/293582745-7caae962-31fa-4c66-bbcb-b1b462189dca.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQyMjg2ODQsIm5iZiI6MTcwNDIyODM4NCwicGF0aCI6Ii8xNjk0Njc4Ni8yOTM1ODI3NDUtN2NhYWU5NjItMzFmYS00YzY2LWJiY2ItYjFiNDYyMTg5ZGNhLlBORz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTAyVDIwNDYyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA3ZTczOWNlOThjNzA0YTYyNTlmNDdkYTNhMDYwMWY0MjQ3NzZhYjA1MmI2ZmI2OGE1MmRiZGU3YWIzZTc2ZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ztjQxDYBFruU-O19A2BXxKEprfzsGTWBle1rrSaa6Ek)

I think q>0 was intended for large meshes > maxIndexableCoords:

https://github.com/x3dom/x3dom/blob/a0601f816d72b16309976ad5657e7212e4c26a5e/src/util/Utils.js#L17
https://github.com/x3dom/x3dom/blob/a0601f816d72b16309976ad5657e7212e4c26a5e/src/gfx_webgl.js#L199

But with webGL2 this probably now rarely happens.

It is possible that tangent and binormalBuffer were added relatively late, without considering the case of q>0 but I do not know.

https://github.com/x3dom/x3dom/blob/a0601f816d72b16309976ad5657e7212e4c26a5e/src/Internals.js#L68

> https://github.com/x3dom/x3dom/blob/a0601f816d72b16309976ad5657e7212e4c26a5e/src/gfx_webgl.js#L842-L846
> 
> 
> And, does it should be `gl.enableVertexAttribArray( sp.binormal); `?

I agree, it probably should be `sp.binormal`. Looks like a copy and paste oversight. But I do not know how to generate a test case (IFS with binormals).

**Comment by microaaron** [comment]
> https://github.com/x3dom/x3dom/blob/a0601f816d72b16309976ad5657e7212e4c26a5e/src/gfx_webgl.js#L199
> 
> But with webGL2 this probably now rarely happens.

Yes, this is also the limit of Array. https://262.ecma-international.org/5.1/#sec-15.4.2.2

**Comment by tsturm** [comment]
Hi @microaaron,

I've opened a Pull Request (https://github.com/x3dom/x3dom/pull/1328) that updates all our devDependencies to address the deprecation warning of punycode in Node.js 21.

Additionally, I've fixed the tangent typo in the following commit: 6df4e397aeea7c90534bb7dd3361f6d891cae9bc

**Comment by microaaron** [comment]
Thank you.

**Comment by microaaron** [comment]
https://github.com/x3dom/x3dom/blob/55f53d110de170efe713dab209cac7f6062640a7/build/jsdoc/jsdoc-template/publish.js#L5

Does it work? Have you tried building it? Does it need to be changed to```taffy = require('@jsdoc/salty').taffy```?
